Interrupt Progress + Progress Action Improvements

chairbender opened the pull request:
### Purpose
Fixes #2204 

Refactors progress actions to provide a more usable and flexible API, decoupling the logic of progress from the UI components. It should now be capable of supporting a wider range of use cases without having to rework the internals of progress actions.
Adds an ActionTarget class which should be a useful way to abstract out the differences between actions which target tiles and actions which target specific objects. This is intended to be used for more than just progress actions, for example some interaction logic and other systems which involve targeting something could benefit from not having to distinguish between tile targets and object targets.

chairbender committed:
rename ProgressAction to OldProgressAction to begin refactor to support more extensive progress logic

chairbender committed:
Refactor progress actions to support more flexibility and ease of use.

chairbender committed:
Interrupt weld actions if welder turns off

chairbender committed:
Add ActionTarget to abstract differences between tile / object targets.
Fix various bugs in new progress action system.

chairbender committed:
Actually call finish progress action, prevent start progress message playing
multiple times when progress is already started.

fomalsd commented on UnityProject/Assets/Scripts/UI/ProgressBar/StandardProgressAction.cs:
```diff
@@ -0,0 +1,279 @@
+			return false;
+		}
+
+		//is this cross matrix? if so, don't start progress if matrix is moving
```
I assume this is temporary?

chairbender replied:
Not really, there's basically no reason I can think of to allow progress to be started when a crossed matrix is moving.

There is a bug here where it doesn't check if the player's own matrix is moving, though.

fomalsd reviewed (approved):
Looking very good code wise! not able to playtest though

chairbender committed:
Merge branch 'develop' into progress-interrupt

chairbender committed:
Check both source and target matrix for cross matrix standard progress actions,
don't allow if either is moving or moves / rotates.
Add defensive inventory add logic to help prevent issue where inventory population
fails due to missing reference issue and player doesn't spawn correctly and
spawn menu stays open. Doesn't fix reference issue but fixes the spawn being botched.
